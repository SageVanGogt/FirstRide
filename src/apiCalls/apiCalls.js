import { geoKey } from './../apiKey';

const signinUser = async (user) => {
  const url = 'http://localhost:3000/api/users/';
  const init = {
    method: "POST",
    body: JSON.stringify(user),
    headers: {
      "Content-Type": "application/json"
    }
  };
  const response = await fetch(url, init);
  const profile = await response.json();
  return profile;
}; 

const signupUser = async (user) => {
  const url = 'http://localhost:3000/api/users/new';
  const body = {
    method: "POST",
    body: JSON.stringify(user),
    headers: {
      "Content-Type": "application/json"
    }
  };
  const response = await fetch(url, body);
  const userRecieved = await response.json();
  return userRecieved;
};

const fetchDestination = async (name) => {
  const url = `http://localhost:3000/api/locations/${name}`;
  const response = await fetch(url);
  const destination = await response.json();
  return destination;
};

const fetchRides = async (destination) => {
  const url = `http://localhost:3000/api/rides/get/${destination}`;
  const body = {
    method: 'POST', 
    headers: {
      "Content-Type": "application/json"
    }
  };
  const response = await fetch(url, body);
  const rides = await response.json();
  return rides;
};

const fetchPickups = async (location) => {
  const url = `http://localhost:3000/api/pickup/get/${location}`;
  const body = {
    method: "POST", 
    headers: {
      "Content-Type": "application/json"
    }
  };
  const response = await fetch(url, body);
  const pickups = await response.json();
  return pickups;
};

const fetchGeocode = async (address) => {
  const url = `https://maps.googleapis.com/maps/api/geocode/json?address=${address}&key=${geoKey}`;  
  try {
    const response = await fetch(url);
    const locationData = await response.json();
    if (!locationData) {
      return new Error('error')
    }
    return locationData;
  } catch (error) {
    throw error;
  }
}; 

const submitNewRide = async (ride) => {
  const url = `http://localhost:3000/api/rides/new`;
  const body = {
    method: 'POST',
    body: JSON.stringify(ride),
    headers: {
      "Content-Type": "application/json"        
    }
  };
  const response = await fetch(url, body);
  const rideId = await response.json();
  return rideId;
};

const submitNewPickup = async (location) => {
  const url = `http://localhost:3000/api/pickup/new`;
  const body = {
    method: 'POST',
    body: JSON.stringify(location),
    headers: {
      "Content-Type": "application/json"        
    }
  };
  const response = await fetch(url, body);
  const pickupId = await response.json();
  return pickupId;
};

const fetchRidesPassengers = async (locationId) => {
  const url = `http://localhost:3000/api/rides_passengers/get/passengers/${locationId}`;
  const body = {
    method: 'POST', 
    headers: {
      'Content-Type': 'application/json'
    }
  };
  const response = await fetch(url, body);
  const passengers = await response.json();
  return passengers;
};

const postRidesPassengers = async (newPassenger) => {
  const url = `http://localhost:3000/api/rides_passengers/new`;   
  const body = {
    method: "POST",
    body: JSON.stringify(newPassenger),
    headers: {
      'Content-Type': 'application/json'
    }
  };
  const response = await fetch(url, body);
  const ridePassenger = await response.json();
  return ridePassenger;
};

const postNewProfile = async (profile) => {
  const url = `http://localhost:3000/api/profiles/new`;
  const body = {
    method: "POST",
    body: JSON.stringify(profile),
    headers: {
      "Content-Type": "application/json"
    }
  };
  const response = await fetch(url, body);
  const profileInfo = await response.json();
  return profileInfo;
};

const removePassengerRide = async (rideId, passengerId, destinationId) => {
  const url = `http://localhost:3000/api/rides/${rideId}/passengers/${passengerId}/destination/${destinationId}`; 
  const body = {
    method: "DELETE",
    headers: {
      "Content-Type": "application/json"
    }
  };
  const response = await fetch(url, body);
  const deletedId = await response.json();
  return deletedId;
};

const fetchUserRides = async (userId) => {
  const url = `http://localhost:3000/api/rides_passengers/get/rides/${userId}`;
  const body = {
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    }
  };
  const response = await fetch(url, body);
  const rides = response.json();
  return rides;
};

const fetchRidesFromUser = async (rideId) => {
  const url = `http://localhost:3000/api/rides/${rideId}/get/user/`;
  const body = {
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    }
  };
  const response = await fetch(url, body);
  const ridesInfo = await response.json();
  return ridesInfo;
};

const reverseGeoCode = async (lat, lng) => {
  const url = `https://maps.googleapis.com/maps/api/geocode/json?latlng=${lat},${lng}&key=${geoKey}`;
  const response = await fetch(url);
  const geoData = await response.json();
  return geoData;
}

export {
  signinUser,
  signupUser,
  fetchDestination,
  fetchRides,
  fetchPickups,
  fetchGeocode,
  submitNewRide,
  submitNewPickup,
  fetchRidesPassengers,
  postRidesPassengers,
  postNewProfile,
  removePassengerRide,
  fetchUserRides,
  fetchRidesFromUser,
  reverseGeoCode
};