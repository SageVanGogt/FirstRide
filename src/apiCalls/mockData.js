export const mockUser = {
  id:1,
  user_id:"456",
  user_img:"http://us.cdn281.fansshare.com/photos/channingtatum/bfhgoi-ccaagvrc-large-fat-44506770.jpg",
  bio:"I like to eat",
  rating:"Solid 7"
}