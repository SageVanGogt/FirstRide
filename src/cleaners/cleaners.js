export const geocodeCleaner = (data) => {
  const results = data.results[0];
  const location = results.geometry.location;
  return location;
}

export const seatsRemainingUpdate = (rides, passengers) => {
  const updatedRides = rides.map(ride => {
    const ridesPassengers = passengers.filter(rider => rider.ride_id === ride.id);
    const totalPassengers = ridesPassengers.length;
    ride.seats_remaining = parseInt(ride.seats_remaining) - totalPassengers;
    return ride;
  })
  
  return updatedRides;
}

export const latLngToAddress = (data) => {
  let address = data.results[0].formatted_address;
  return address;
}