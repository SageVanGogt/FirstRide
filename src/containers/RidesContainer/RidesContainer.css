.ride-list {
  width: 300px;
}

.ride-item {
  padding: 10px 0;
}

.ride-item:hover {
  background-color: pink;
}

.ride-page {
  margin-top: 72px;
  height: calc(100vh - 90px);
  background-color: grey;
  width: 300px;
  display: flex;
  overflow: scroll;
}

.offer-btn {
  display: flex;
  justify-content: center;
  align-items: center;
  position: fixed;
  bottom: 0;
  width: 300px;
  font-size: 1.5rem;
  font-family: 'Roboto', sans-serif;  
  background-color: #404040;
  text-decoration: none;
  height: 60px;
  color: white;
  border: none;
  cursor:pointer;
}

.current-location-form { 
  width: 300px;
  display: flex;
  height: 50px;
}

.current-location-form input {
  border: none;
}

.street-input {
 width: 125px;
}

.city-input {
  width: 85px;
} 

.state-input {
  width: 33px;
}