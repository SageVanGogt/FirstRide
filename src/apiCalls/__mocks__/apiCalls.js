import * as MOCK from './../mockData';

const signinUser = jest.fn().mockImplementation(() => 
  Promise.resolve({
    user: [{
    "id":1,
    "user_id":"456",
    "user_img":"http://us.cdn281.fansshare.com/photos/channingtatum/bfhgoi-ccaagvrc-large-fat-44506770.jpg",
    "bio":"I like to eat",
    "rating":"Solid 7"
    }]
  })
);

const signupUser = jest.fn().mockImplementation(() => Promise.resolve({
  "id": 1
}));

const fetchDestination = jest.fn().mockImplementation(() => Promise.resolve({
  "id": 1,
  "location_name": "Red Rocks",
  "location_lat_lng": "12341, 123412"
}));

const fetchRides = jest.fn().mockImplementation(() => 
Promise.resolve(MOCK.mockRides));

const fetchPickups = jest.fn().mockImplementation(() => 
Promise.resolve(MOCK.mockPickups));

const fetchGeocode = jest.fn().mockImplementation(() => 
Promise.resolve(MOCK.mockGeoInfo));

const submitNewRide = jest.fn().mockImplementation(() => 
Promise.resolve({
  id: 1
}));

const submitNewPickup = jest.fn().mockImplementation(() => 
Promise.resolve({
  id: 1
}));

const postRidesPassengers = jest.fn().mockImplementation(() => 
Promise.resolve({
  id: 1
}));

export {
  signinUser,
  signupUser,
  fetchDestination,
  fetchRides,
  fetchPickups,
  fetchGeocode,
  submitNewPickup,
  submitNewRide,
  postRidesPassengers
}