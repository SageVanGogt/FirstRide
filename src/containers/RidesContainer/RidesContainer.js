import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import MapContainer from './../MapContainer/MapContainer';
import * as API from './../../apiCalls/apiCalls';
import * as actions from './../../actions/rides';
export class RidesContainer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      street: '',
      city: '',
      state: ''
    };
  }

  componentDidUpdate = (prevProps) => {
    if (prevProps.destination.id !== this.props.destination.id) {
      this.loadRides();
    }
  }

  loadRides = async () => {
    const { setRides, destination } = this.props;
    const response = await API.fetchRides(destination.id);
    await setRides(response.rides);
  }

  handleChange = (event) => {
    const { name, value } = event.target;
    this.setState({
      [name]: value
    });
  }

  formatAddress = () => {
    const street = this.state.street.replace(' ', '+');
    const city = this.state.city.replace(' ', '+');
    const state = this.state.state.replace(' ', '+');
    return (
      `${street},+${city},+${state}`
    );
  }

  render() {
    return (
      <div>
        <form action="submit" onSubmit={this.handleSubmit}>
          <input 
            type="text" 
            name="street" 
            onChange={this.handleChange} 
            placeholder="street"/>
          <input 
            type="text" 
            name="city" 
            onChange={this.handleChange} 
            placeholder="city"/>
          <input 
            type="text" 
            name="state" 
            onChange={this.handleChange} 
            placeholder=""/>
          <input type="state"/>
        </form>
        <MapContainer />
      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  destination: state.destination
})

export const mapDispatchToProps = (dispatch) => ({
  setRides: (rides) => dispatch(actions.addRides(rides))
})

export default connect(mapStateToProps, mapDispatchToProps)(RidesContainer);