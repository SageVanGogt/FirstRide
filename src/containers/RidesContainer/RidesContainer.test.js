import React from 'react';
import { shallow } from 'enzyme';
import { RidesContainer, mapStateToProps, mapDispatchToProps } from './RidesContainer';
import * as MOCK from './../../apiCalls/mockData';
import * as API from './../../apiCalls/apiCalls';
import * as cleaner from './../../cleaners/cleaners';

jest.mock('./../../apiCalls/apiCalls');

describe('RidesContainer', () => {
  let wrapper;
  let mockDestination;
  let mockSetRides;
  let mockSetLocation;
  let mockRides;
  let mockUser;

  beforeEach(() => {
    mockUser = {
      id: 1
    }
    mockDestination = {
      location_name: 'Red Rocks',
      id: 1
    };
    mockRides = MOCK.mockRides.rides;
    mockSetRides = jest.fn();
    mockSetLocation = jest.fn();
    wrapper = shallow(<RidesContainer 
      rides={mockRides}
      setLocation={mockSetLocation}
      destination={mockDestination}
      setRides={mockSetRides}
      user={mockUser}/>
    );
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  describe('handleChange', () => {
    
    it('should update the correct state with the value of the input', () => {
      let mockEvent = {
        target: {
          value: 'seattle', 
          name: 'city'
        }
      };
      let expected = 'seattle';
      wrapper.instance().handleChange(mockEvent);
      let actual = wrapper.state('city');
      
      expect(actual).toEqual(expected);
    })
  })

  describe('handleSubmit', () => {
    let mockEvent;

    beforeEach(() => {
      mockEvent = {
        preventDefault: jest.fn()
      };
      wrapper.setState({
        street: '2600 fairview',
        city: 'seattle', 
        state: 'WA'
      })
    });

    it('should call fetchGeocode with the correct params', async () => {
      let expected = '2600+fairview,+seattle,+WA';
      
      await wrapper.instance().handleSubmit(mockEvent);

      expect(API.fetchGeocode).toHaveBeenCalledWith(expected);
    });

    it('should call setLocation with the correct params', async () => {
      let expected = {
        lat: 39.7594866,
        lng: -104.9994026
      };
      await wrapper.instance().handleSubmit(mockEvent);

      expect(mockSetLocation).toHaveBeenCalledWith(expected);
    });
    
    it('should call geocodeCleaner with the correct params', async () => {
      let expected = MOCK.mockGeoInfo;
      
      cleaner.geocodeCleaner = jest.fn();
      await wrapper.instance().handleSubmit(mockEvent);

      expect(cleaner.geocodeCleaner).toHaveBeenCalledWith(expected);
    });
  });

  describe('formatAddress', () => {

    it('should return an address with no spaces', () => {
      let expected = '2600+fairview,+seattle,+WA'
      wrapper.setState({
        street: '2600 fairview',
        city: 'seattle', 
        state: 'WA'
      })
      let actual = wrapper.instance().formatAddress();
      
      expect(actual).toEqual(expected);
    })
  });

  describe('loadRides', () => {
    it('should call setRides with correct params', async () => {
      const expected = MOCK.mockUpdatedRides.rides
      await wrapper.instance().loadRides()
      expect(mockSetRides).toHaveBeenCalledWith(expected)
    })

    it('should call fetchRides with the correct params', async () => {
      await wrapper.instance().loadRides();
      expect(API.fetchRides).toHaveBeenCalledWith(mockDestination.id)
    })

    it('should call fetchRidesPassengers with correct params', async () => {
      const expected = 1
      await wrapper.instance().loadRides()
      expect(API.fetchRidesPassengers).toHaveBeenCalledWith(expected)
    })
  })

  describe('mapStateToProps', () => {
    
    it('should return a state with the destination prop', () => {
      let mockState = {
        destination: 'Red Rocks',
        user: {},
        rides: []
      };
      let mappedProps = mapStateToProps(mockState);
      let actual = mappedProps.destination;
      let expected = 'Red Rocks';

      expect(actual).toEqual(expected);
    })

    it('should return a state with the rides prop', () => {
      let mockState = {
        destination: 'Red Rocks',
        user: {},
        rides: [{}, {}]
      };
      let mappedProps = mapStateToProps(mockState);
      let actual = mappedProps.rides;
      let expected = [{}, {}];

      expect(actual).toEqual(expected);
    })

    it('should return a state with the user prop', () => {
      let mockState = {
        destination: 'Red Rocks',
        user: {id: 1},
        rides: [{}, {}]
      };
      let mappedProps = mapStateToProps(mockState);
      let actual = mappedProps.user;
      let expected = {id: 1};

      expect(actual).toEqual(expected);
    })
  })

  describe('mapDispatchToProps', () => {

    it('should dispatch setRides with the correct params', () => {
      let mockDispatch = jest.fn();
      let mappedProps = mapDispatchToProps(mockDispatch);
      let expected = {
        type: "ADD_RIDES",
        rides: MOCK.mockRides.rides
      }
      mappedProps.setRides(MOCK.mockRides.rides);

      expect(mockDispatch).toBeCalledWith(expected)
    })

    it('should dispatch setLocation with the correct params', () => {
      let mockDispatch = jest.fn();
      let mappedProps = mapDispatchToProps(mockDispatch);
      let expected = {
        type: "ADD_CURR_LOCATION",
        location: {
          lat: 39.7594866,
          lng: -104.9994026
        }
      };
      mappedProps.setLocation(expected.location);

      expect(mockDispatch).toBeCalledWith(expected)
    })
  })

  describe('submitRideSignup', () => {
    
    it('should call postRidesPassengers with the corerct params', async () => {
      let expected = {
        ride_id: 1,
        passenger_id: 1,
        location_id: 1
      }
      await wrapper.instance().submitRideSignup(1);

      expect(API.postRidesPassengers).toHaveBeenCalledWith(expected);
    })
  })
})