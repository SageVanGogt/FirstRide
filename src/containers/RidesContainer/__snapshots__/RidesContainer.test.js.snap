// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RidesContainer should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RidesContainer
    destination={
        Object {
            "id": 1,
            "location_name": "Red Rocks",
          }
    }
    setLocation={[Function]}
    setRides={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <form
          action="submit"
          onSubmit={[Function]}
>
          <input
                    name="street"
                    onChange={[Function]}
                    placeholder="street"
                    type="text"
          />
          <input
                    name="city"
                    onChange={[Function]}
                    placeholder="city"
                    type="text"
          />
          <input
                    name="state"
                    onChange={[Function]}
                    placeholder="state"
                    type="text"
          />
          <input
                    type="submit"
          />
</form>,
        <Connect(MapContainer) />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "action": "submit",
          "children": Array [
            <input
              name="street"
              onChange={[Function]}
              placeholder="street"
              type="text"
/>,
            <input
              name="city"
              onChange={[Function]}
              placeholder="city"
              type="text"
/>,
            <input
              name="state"
              onChange={[Function]}
              placeholder="state"
              type="text"
/>,
            <input
              type="submit"
/>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "street",
              "onChange": [Function],
              "placeholder": "street",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "city",
              "onChange": [Function],
              "placeholder": "city",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "state",
              "onChange": [Function],
              "placeholder": "state",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "type": "submit",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <form
            action="submit"
            onSubmit={[Function]}
>
            <input
                        name="street"
                        onChange={[Function]}
                        placeholder="street"
                        type="text"
            />
            <input
                        name="city"
                        onChange={[Function]}
                        placeholder="city"
                        type="text"
            />
            <input
                        name="state"
                        onChange={[Function]}
                        placeholder="state"
                        type="text"
            />
            <input
                        type="submit"
            />
</form>,
          <Connect(MapContainer) />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "action": "submit",
            "children": Array [
              <input
                name="street"
                onChange={[Function]}
                placeholder="street"
                type="text"
/>,
              <input
                name="city"
                onChange={[Function]}
                placeholder="city"
                type="text"
/>,
              <input
                name="state"
                onChange={[Function]}
                placeholder="state"
                type="text"
/>,
              <input
                type="submit"
/>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "street",
                "onChange": [Function],
                "placeholder": "street",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "city",
                "onChange": [Function],
                "placeholder": "city",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "state",
                "onChange": [Function],
                "placeholder": "state",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "type": "submit",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
